lib/ansible/plugins/loader.py

```diff
@@ -676,11 +805,18 @@ def _load_plugin_filter():
     return filters
 
 
+def _configure_collection_loader():
+    if not any((isinstance(l, AnsibleCollectionLoader) for l in sys.meta_path)):
+        sys.meta_path.insert(0, AnsibleCollectionLoader())
+
+
 # TODO: All of the following is initialization code   It should be moved inside of an initialization
 # function which is called at some point early in the ansible and ansible-playbook CLI startup.
 
 _PLUGIN_FILTERS = _load_plugin_filter()
 
+_configure_collection_loader()
```

Comment:
Instead of doing this at module load time, it may be more predictable to just call it early in the app entry points (possibly main or somewhere in cli/ ?). The ordering of when/how the ansible.* modules gets loaded can easily change and change collection loading behavior. Those sort of bugs can be very hard to debug.

Folks importing ansible code in other apps would need to add the call to _configure_collection_loader() to their startup if they want the collection loader.

Reply:
I think the places it would need to be called explicitly are:
 - cli app startup (bin/ansible? ansible.cli.*?)
 - ansiballZ startup
 - worker process startup? (not require, but possibly useful)

For worker process, I dont think they _have_ to resetup the collection loader,
but if they did create a new AnsibleCollectionLoader (and replace the one in sys.metapath
for the process if needed), that would help isolate workers from potentially altering
the controllers collection loader.

For ansiballz, with an explicit _configure_collection_loader() and AnsibleCollectionLoader creation,
it could use a different config and potentially offer a place to add/tweak python module loading on the remote
side. However, the interaction between PluginLoader, module_common, and ansiballZ imports is already
very complicated, unless using CollectionLoader could simplify that it's probably not worth changing.